src/librustdoc/passes/collect_intra_doc_links.rs

```diff
@@ -837,7 +839,37 @@ impl<'a, 'tcx> DocFolder for LinkCollector<'a, 'tcx> {
             Some(if parent_kind == DefKind::Impl { parent } else { item.def_id.expect_real() })
         } else {
             Some(item.def_id.expect_real())
-        };
+        }
+        .map(|curr_id| match self.cx.tcx.def_kind(curr_id) {
+            DefKind::Impl => {
+                match curr_id
+                    .as_local()
+                    .map(|local_id| self.cx.tcx.hir().item(rustc_hir::ItemId { def_id: local_id }))
+                {
+                    Some(rustc_hir::Item {
+                        kind:
+                            rustc_hir::ItemKind::Impl(rustc_hir::Impl {
+                                self_ty:
+                                    rustc_hir::Ty {
+                                        kind:
+                                            rustc_hir::TyKind::Path(rustc_hir::QPath::Resolved(
+                                                _,
+                                                rustc_hir::Path {
+                                                    res: rustc_hir::def::Res::Def(_, def_id),
+                                                    ..
+                                                },
+                                            )),
+                                        ..
+                                    },
+                                ..
+                            }),
+                        ..
+                    }) => *def_id,
+                    _ => curr_id,
+                }
+            }
+            _ => curr_id,
```

Comment:
This is all wrong, it will break for impls from foreign crates. Use the logic below from `self_name` instead.

(This is what I removed initially in https://github.com/rust-lang/rust/pull/76467.)